### ISSUE TITLE
any chance to see a changelog for the releases some day or at least meaningful commit messages?

### ISSUE THREAD

**Issue by stohrendorf** [dtschump/CImg#316]
I like CImg, because it's just a plain header file, but I have a huge problem with the missing change descriptions. I like having up-to-date libs in my projects, but CImg is unfortunately a blind spot, because updating it is a game of luck as things may break or things may be fixed, or new features are suddenly there without notice or vanish without notice, which is sad. One time I bumped CImg certain key parts of my project broke because an image format was removed without further notice, and I had to spend quite some time finding out which was the highest version if CImg supporting that particular image format, and finally ended up writing my own loader for it. It takes a very long time to step through every commit and understand how CImg works internally to see which changes are actually useful to me and which changes require me to take action. There are at least a hundred of commits for every release, and I need to read every single one to see what's changed. You could make quite a few people's life easier if you could spend a few seconds and write proper commit messages, even if you don't want to write release notes (but writing release notes becomes a lot easier if you have a list of good commit messages at hand).

Cheers.

**Comment by dtschump** [comment]
Hello Steffen,
You are absolutely right. And you are not the first to tell me that.

The fact is that I actually (try to) manage two open-source projects, [CImg](http://cimg.eu/) and [G'MIC](https://gmic.eu/), in addition to my work as a researcher (not to mention my family life), and when I code for my open-source projects, I don't take the time to put meaningful commit messages, mainly because I find that I'm already not moving fast enough.
This is a real problem. **I always code in a hurry**,  because I've the impression I have dozens of other things to do at the same time.

So my solution is to focus on code progress rather than documentation. This fact is also visible in the CImg reference documentation: many functions are not documented as they should be, details are missing, and sometimes the parameter names are even wrong. At the moment, I work much more on G'MIC than on CImg, but this global documentation problem is the same (maybe even worse) for G'MIC.

About release notes : at least, I'm trying to write a changelog for G'MIC, and it usually contains some of the changes I've made in CImg as well (as G'MIC is based on CImg). You can find it here : https://discuss.pixls.us/t/on-the-road-to-3-0/17226
Not of a great help, anyway, for pure CImg users.

At this stage, I don't really have a solution though. 
For the moment, at least both projects are working technically, even without much documentation. I just think we need more involved contributors to get everything right (including the commit messages, but not only :) ).

Cheers,

David.

**Comment by stohrendorf** [comment]
Even if you're in a hurry, you do the commits with a goal in mind, so I think you can write that goal as a one-liner in the commit messages, e.g.
* 54362beceb451a26a414743951bb9ea36e1f75d3 -> fix unresolved merge
* 94427018aca3e01df40f963c9432d5c59d5338ed -> add 2.9.7 pre-release links
* dd6ab6b995e576772b5c740d0eff90d5b11bc240 -> add 2.9.5 release links
* 7b918ee29384e6bcd4b7b105ce098f30e058ab17 -> refactor/simplify/consolidate bool image conversion
* 0cdc54bcfb61bed1a52f37c7596989fab630abac -> don't open internal links in new window, exchange youtube demo video link
* 51804665e3d2e91c8066be95b9bb0b9db5234271 -> clean up documentation files
* 22e8ca938680841773aa8b2999d270894b89a1a9 -> add new documentation

Even while I had to guess some of the intentions, and I had no idea at all about what some commits changed, every message took me less than 20 seconds to come up with and write down, including understanding what changed. This is most certainly only a tiny fraction of the time it takes you to do the actual code changes.

Another problem that arises from not writing proper commit messages is that if I had looked at the commit history before integrating CImg, I would never have considered using it at all, because it simply says "the author doesn't care about what he's doing", and I don't want to use libs where quality is not one of the primary concerns.

**Comment by dtschump** [comment]
> Even if you're in a hurry, you do the commits with a goal in mind, so I think you can write that goal as a one-liner in the commit messages, e.g.

Not at all. Personnally, I do a commit only when I want my code to be archived at some point. One commit is often not related to the addition of a specific feature. That's how I work. Often lot of commits may happen for a single feature. Sometimes, a single commit for several features at the same time. All in all, looking at the commit history in the hope to get an idea of the changelog between two versions is actually not a good idea. What I should do is preferably write a real changelog somewhere.

> Another problem that arises from not writing proper commit messages is that if I had looked at the commit history before integrating CImg, I would never have considered using it at all, because it simply says "the author doesn't care about what he's doing", and I don't want to use libs where quality is not one of the primary concerns.

If it's what you'd think about my care about the lib quality, I think you are wrong.
But I'm not here to change your mind. I'm here to spend the little free time I have to do my best creating a C++ image processing library. Not perfect of course, but really, I can't spend 100% of my time on this (not even 10%). Unless there are motivated contributors to improve the library, it can't be perfect as long as I'm alone working on it.

**Comment by stohrendorf** [comment]
> Not at all. Personnally, I do a commit only when I want my code to be archived at some point. One commit is often not related to the addition of a specific feature. That's how I work. Often lot of commits may happen for a single feature. Sometimes, a single commit for several features at the same time. All in all, looking at the commit history in the hope to get an idea of the changelog between two versions is actually not a good idea. What I should do is preferably write a real changelog somewhere.

I do work the same, but I have a bit more time at hand, yet I end up with the same result most of the time. When I do a release, I look at the commit messages that happened since the last release, and basically copy over the most relevant commits. It also helps cleaning up my mind as I don't have to remember what changes are introduced, I do have my commit messages I can look at.

> If it's what you'd think about my care about the lib quality, I think you are wrong.

I'm not thinking that, sorry if you got the impression. But reading how you think about the unnecessity of commit messages, _that's_ starting to change my mind.

> But I'm not here to change your mind. I'm here to spend the little free time I have to do my best creating a C++ image processing library. Not perfect of course, but really, I can't spend 100% of my time on this (not even 10%). Unless there are motivated contributors to improve the library, it can't be perfect as long as I'm alone working on it.

Contributors won't help you with your own commit messages. You said this is a core lib used in gmic, and you're doing what's needed. Given that gmic is fairly popular and in rapid development, this means that you will be the main developer for CImg as long as you're the main developer on gmic. This in turn means most of the commits will be done by you, because (from my own experience) _code_ contributions to core libraries are fairly low compared to the contributions from the actual core developers resolving the reported issues. Thus, nearly all commit messages will be a `.` even while seeing a contributor from time to time.

What's more, if you get a contributor, you also need to inform the contributor what you're working on, so he can judge whether your changes affect his changes. You're basically saying "I do the work, give me your work, I don't care if I break your work, and while you're working on it, fix my commit messages, but I don't have any time at all to care about you, just fix your code whenever I break it, thank you." You may have a different mind, but your statements are communicating exactly that, sorry.

Also, maybe you misunderstood me, but I'm not asking for a perfect library, I'm asking for the existence of commit messages. Say you can spend 5% of your time working on this project, I highly doubt you would spend the other 95% on writing somewhat useful commit messages and not getting anything else done. It's just a matter of habit, not a matter of work. Because I just can't imagine you can't even spend 3 minutes _a day_ to add a few helpful commit messages. I gave you some examples that would be enough and actually describe what happened in these commits.

CImg is a good library, it does what it's intended to do, but as it seems, you don't even have time to think about what I'm trying to give you advice about, and I'm sorry to see this happen.
